File: latticeai/core/csrf.py
"""Origin/Referer guard for cookie-authenticated state changes.

The ``session_token`` cookie is ``SameSite=Lax``, which stops cross-site
*top-level GET-like* navigations from carrying it but does **not** stop a
cross-site ``POST``/``PUT``/``PATCH``/``DELETE`` issued by script from a page
the browser considers a different site in every configuration we ship (Lax
blocks non-GET cross-site sends in current browsers, but the cookie is also
the only credential a browser attaches automatically, and Lax is not a
security boundary we should rely on alone). Every mutating endpoint accepted
that cookie with no second check, so a non-loopback deployment was one
malicious page away from forged state changes.

This module is the *decision*; :mod:`latticeai.runtime.web_runtime` is the
wiring. Keeping the two apart means the policy is testable without an ASGI
app, and the middleware stays small enough to audit.

Threat model, stated so the exemptions are checkable rather than vibes:

* **Only ambient credentials are CSRF-able.** ``Authorization: Bearer`` is set
  by the caller's own code; an attacker's page cannot add it to a cross-site
  request without a CORS preflight that our ``CORSMiddleware`` allowlist
  rejects. Bearer-authenticated requests are therefore exempt.
* **No cookie, nothing to forge.** A request without ``session_token`` cannot
  be authenticated by cookie, so it is not this guard's business.
* **``Origin`` is attacker-honest.** A browser sets it; a page cannot lie about
  it. ``Referer`` is the fallback for the handful of clients that omit
  ``Origin``.
* **Neither header present** means the caller is not a browser (curl, the CLI,
  the desktop shell, the VS Code extension). That is trusted only while the
  server is bound to loopback, where "not a browser" and "already on this
  machine" coincide. On a reachable bind it fails closed.
* **Behind a proxy, ``Host`` is not the front door.** The desktop gateway
  replaces it with the worker's internal authority, so the ``Origin == Host``
  fallback below compared a browser's real origin against a port the browser
  has never heard of and refused every write. The front door's own name comes
  from :mod:`latticeai.core.http_origin`, which believes ``X-Forwarded-Host``
  only from a peer that may forward.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import (
    Any,
    Awaitable,
    Callable,
    Iterable,
    List,
    MutableMapping,
    Optional,
    Sequence,
    Tuple,
)
from urllib.parse import urlsplit

from latticeai.core.http_origin import (
    FORWARDED_HOST_HEADER,
    effective_host,
)

__all__ = [
    "CSRFDecision",
    "CSRFOriginPolicy",
    "CSRFOriginGuardMiddleware",
    "SAFE_METHODS",
    "SESSION_COOKIE_NAME",
    "normalize_origin",
]

# The ASGI contract, spelled out locally so the policy stays importable (and
# unit-testable) without pulling in a web framework.
Scope = MutableMapping[str, Any]
Message = MutableMapping[str, Any]
Receive = Callable[[], Awaitable[Message]]
Send = Callable[[Message], Awaitable[None]]
ASGIApp = Callable[[Scope, Receive, Send], Awaitable[None]]

# Methods that must not change state. Everything else is guarded.
SAFE_METHODS = frozenset({"GET", "HEAD", "OPTIONS", "TRACE"})

SESSION_COOKIE_NAME = "session_token"

_DEFAULT_PORTS = {"http": 80, "https": 443, "ws": 80, "wss": 443}

_DENIED_DETAIL = (
    "요청 출처를 확인할 수 없어 거부했습니다. "
    "다른 사이트에서 보낸 요청일 수 있습니다."
)


def normalize_origin(value: Optional[str]) -> Optional[Tuple[str, str, Optional[int]]]:
    """``"HTTP://Localhost:80/x"`` → ``("http", "localhost", None)``.

    Returns ``None`` for anything that is not a usable origin (empty, the
    literal ``"null"`` that sandboxed iframes and ``file://`` pages send, or a
    value with no host). ``"null"`` is deliberately *not* normalized into a
    matchable origin: it is what an opaque origin sends, and opaque origins are
    exactly the ones we must not trust.
    """
    if not value:
        return None
    candidate = value.strip()
    if not candidate or candidate.lower() == "null":
        return None
    if "//" not in candidate:
        # A bare authority ("example.com:4825") — the Host header shape.
        candidate = "//" + candidate
    parts = urlsplit(candidate)
    host = (parts.hostname or "").lower()
    if not host:
        return None
    scheme = (parts.scheme or "").lower()
    try:
        port = parts.port
    except ValueError:
        # Malformed port ("example.com:notaport"): unusable, so untrusted.
        return None
    if port is not None and _DEFAULT_PORTS.get(scheme) == port:
        port = None
    return (scheme, host, port)


def _same_site(left: Tuple[str, str, Optional[int]], right: Tuple[str, str, Optional[int]]) -> bool:
    """Host+port equality, ignoring scheme when one side does not carry one.

    The ``Host`` header has no scheme, and a TLS-terminating reverse proxy
    speaks ``http`` to us while the browser reports ``https``. Comparing the
    authority is what makes "this page came from this server" answerable in
    both deployments; the scheme is only compared when both sides state one.
    """
    if left[1] != right[1] or left[2] != right[2]:
        return False
    if left[0] and right[0]:
        return left[0] == right[0]
    return True


@dataclass(frozen=True)
class CSRFDecision:
    """Why a request was allowed or refused — the reason is the audit trail."""

    allowed: bool
    reason: str


class CSRFOriginPolicy:
    """Decide whether one request may change state with a cookie credential."""

    def __init__(
        self,
        *,
        trusted_origins: Iterable[str] = (),
        server_host: str = "127.0.0.1",
        server_port: int = 4825,
        bind_is_loopback: bool = True,
    ) -> None:
        self.bind_is_loopback = bool(bind_is_loopback)
        self._trusted: List[Tuple[str, str, Optional[int]]] = []
        for origin in self._default_origins(server_host, server_port):
            self._add(origin)
        for origin in trusted_origins:
            self._add(origin)

    @staticmethod
    def _default_origins(server_host: str, server_port: int) -> List[str]:
        """The server's own origin plus loopback, in both schemes."""
        hosts = [server_host, "localhost", "127.0.0.1", "[::1]"]
        origins: List[str] = []
        for host in hosts:
            if not host:
                continue
            for scheme in ("http", "https"):
                origins.append(f"{scheme}://{host}:{server_port}")
        return origins

    def _add(self, origin: str) -> None:
        normalized = normalize_origin(origin)
        if normalized is not None and normalized not in self._trusted:
            self._trusted.append(normalized)

    @property
    def trusted_origins(self) -> Sequence[Tuple[str, str, Optional[int]]]:
        return tuple(self._trusted)

    def _origin_is_trusted(
        self,
        origin: Tuple[str, str, Optional[int]],
        host_header: Optional[str],
        forwarded_host: Optional[str] = None,
        peer: Optional[str] = None,
    ) -> bool:
        if any(_same_site(origin, trusted) for trusted in self._trusted):
            return True
        # Same-origin by the request's own Host. A browser sets Host from the
        # URL it is fetching, so `Origin == Host` can only be produced by a
        # page this server actually served — which is precisely "not cross
        # site". This is what keeps reverse-proxied hostnames working without
        # every operator having to enumerate them.
        #
        # "The request's own Host" is the *front door's*, not the internal one a
        # proxy hop rewrote it to, which is why the forwarded authority is
        # resolved first (and only from a peer that may forward).
        own = normalize_origin(
            effective_host(host=host_header, forwarded_host=forwarded_host, peer=peer)
        )
        return own is not None and _same_site(origin, own)

    def evaluate(
        self,
        *,
        method: str,
        origin: Optional[str],
        referer: Optional[str],
        host: Optional[str],
        cookie_header: Optional[str],
        authorization: Optional[str],
        forwarded_host: Optional[str] = None,
        peer: Optional[str] = None,
    ) -> CSRFDecision:
        """Allow/deny one request. Pure: no I/O, no app state."""
        if method.upper() in SAFE_METHODS:
            return CSRFDecision(True, "safe-method")
        if (authorization or "").strip().lower().startswith("bearer "):
            # Not ambient: a cross-site page cannot attach this header.
            return CSRFDecision(True, "bearer-auth")
        if not _has_session_cookie(cookie_header):
            return CSRFDecision(True, "no-session-cookie")

        stated = normalize_origin(origin)
        if stated is None and origin:
            # An opaque origin ("null") explicitly claims *untrusted* provenance.
            return CSRFDecision(False, "opaque-origin")
        if stated is None:
            stated = normalize_origin(referer)
            if stated is None:
                if self.bind_is_loopback:
                    # Non-browser client on the same machine (CLI, desktop
                    # shell, curl). Nothing on the network can reach this bind.
                    return CSRFDecision(True, "no-origin-loopback-bind")
                return CSRFDecision(False, "no-origin-reachable-bind")

        if self._origin_is_trusted(stated, host, forwarded_host, peer):
            return CSRFDecision(True, "same-site-or-trusted-origin")
        return CSRFDecision(False, "cross-site-origin")


def _has_session_cookie(cookie_header: Optional[str]) -> bool:
    """Whether the raw Cookie header carries ``session_token``.

    Parsed by hand rather than through ``http.cookies`` so a malformed pair
    cannot make the whole header unreadable — a request whose cookie jar we
    cannot parse must still be treated as cookie-bearing.
    """
    if not cookie_header:
        return False
    for pair in cookie_header.split(";"):
        name, _, _value = pair.partition("=")
        if name.strip() == SESSION_COOKIE_NAME:
            return True
    return False


class CSRFOriginGuardMiddleware:
    """Pure-ASGI guard: inspects request headers, never touches the response.

    Written against the raw ASGI interface instead of ``BaseHTTPMiddleware``
    on purpose — this app streams (SSE chat, live agent steps), and wrapping
    the response path to make a decision that only needs request headers would
    put a buffering layer in front of every stream for no benefit.
    """

    def __init__(self, app: ASGIApp, *, policy: CSRFOriginPolicy) -> None:
        self.app = app
        self.policy = policy

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if scope.get("type") != "http":
            await self.app(scope, receive, send)
            return
        headers = {
            key.decode("latin-1").lower(): value.decode("latin-1")
            for key, value in scope.get("headers") or []
        }
        # ``client`` is ``(host, port)`` under every ASGI server, and absent
        # under a test transport that does not model a socket.
        client = scope.get("client") or ()
        decision = self.policy.evaluate(
            method=str(scope.get("method") or "GET"),
            origin=headers.get("origin"),
            referer=headers.get("referer"),
            host=headers.get("host"),
            cookie_header=headers.get("cookie"),
            authorization=headers.get("authorization"),
            forwarded_host=headers.get(FORWARDED_HOST_HEADER),
            peer=client[0] if client else None,
        )
        if decision.allowed:
            await self.app(scope, receive, send)
            return
        await _send_forbidden(send, decision.reason)


async def _send_forbidden(send: Send, reason: str) -> None:
    body = json.dumps(
        {"detail": _DENIED_DETAIL, "error": "csrf_origin_rejected", "reason": reason},
        ensure_ascii=False,
    ).encode("utf-8")
    await send(
        {
            "type": "http.response.start",
            "status": 403,
            "headers": [
                (b"content-type", b"application/json; charset=utf-8"),
                (b"content-length", str(len(body)).encode("ascii")),
            ],
        }
    )
    await send({"type": "http.response.body", "body": body})

File: rust/lattice-auth/src/csrf.rs
//! The Origin/Referer guard, ported branch for branch from `core/csrf.py`.
//!
//! Seven branches, in order, and the order *is* the policy — moving branch 3
//! above branch 2 would make bearer clients depend on not sending a cookie,
//! and moving branch 5 above branch 4 would let an opaque origin pass as
//! "no origin at all". The reason label on each decision is the audit trail
//! and rides in the refusal body, so the labels are part of the contract too.
//!
//! Threat model, restated from the original so the exemptions stay checkable:
//! only ambient credentials are forgeable, so `Authorization: Bearer` and
//! "no session cookie" are exempt; `Origin` is attacker-honest because a
//! browser sets it; neither header present means the caller is not a browser,
//! which is trusted only on a loopback bind.

use axum::http::{HeaderMap, Method, StatusCode};
use axum::response::Response;
use serde_json::json;

use crate::cookies::has_session_cookie;
use crate::origin::{effective_host, normalize_origin, same_site, IpNetwork, Origin};
use crate::pyjson::dumps_spaced;
use crate::response::json_response_utf8;

/// Methods that must not change state; everything else is guarded.
pub const SAFE_METHODS: [&str; 4] = ["GET", "HEAD", "OPTIONS", "TRACE"];

/// The refusal text, byte-identical to `_DENIED_DETAIL`.
pub const DENIED_DETAIL: &str =
    "요청 출처를 확인할 수 없어 거부했습니다. 다른 사이트에서 보낸 요청일 수 있습니다.";

/// Why a request was allowed or refused.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CsrfDecision {
    /// Whether the request may proceed.
    pub allowed: bool,
    /// The label that names the branch taken.
    pub reason: &'static str,
}

/// Everything one CSRF decision needs, so the policy stays pure.
#[derive(Debug, Default, Clone)]
pub struct CsrfRequest<'a> {
    /// The HTTP method.
    pub method: &'a str,
    /// The `Origin` header, if any.
    pub origin: Option<&'a str>,
    /// The `Referer` header, if any.
    pub referer: Option<&'a str>,
    /// The `Host` header, if any.
    pub host: Option<&'a str>,
    /// The raw `Cookie` header, if any.
    pub cookie_header: Option<&'a str>,
    /// The `Authorization` header, if any.
    pub authorization: Option<&'a str>,
    /// The `X-Forwarded-Host` header, if any.
    pub forwarded_host: Option<&'a str>,
    /// The direct peer address.
    pub peer: Option<&'a str>,
}

/// Decide whether one request may change state with a cookie credential.
#[derive(Debug, Clone)]
pub struct CsrfOriginPolicy {
    trusted: Vec<Origin>,
    bind_is_loopback: bool,
    trusted_proxies: Vec<IpNetwork>,
}

impl CsrfOriginPolicy {
    /// The server's own origin and loopback are trusted by construction; the
    /// operator's `LATTICEAI_CSRF_TRUSTED_ORIGINS` entries are appended.
    pub fn new(
        trusted_origins: &[String],
        server_host: &str,
        server_port: u16,
        bind_is_loopback: bool,
        trusted_proxies: Vec<IpNetwork>,
    ) -> Self {
        let mut policy = Self {
            trusted: Vec::new(),
            bind_is_loopback,
            trusted_proxies,
        };
        for origin in default_origins(server_host, server_port) {
            policy.add(&origin);
        }
        for origin in trusted_origins {
            policy.add(origin);
        }
        policy
    }

    fn add(&mut self, origin: &str) {
        if let Some(normalized) = normalize_origin(Some(origin)) {
            if !self.trusted.contains(&normalized) {
                self.trusted.push(normalized);
            }
        }
    }

    /// The normalised allowlist, in the order it was built.
    pub fn trusted_origins(&self) -> &[Origin] {
        &self.trusted
    }

    fn origin_is_trusted(
        &self,
        origin: &Origin,
        host_header: Option<&str>,
        forwarded_host: Option<&str>,
        peer: Option<&str>,
    ) -> bool {
        if self
            .trusted
            .iter()
            .any(|trusted| same_site(origin, trusted))
        {
            return true;
        }
        // Same-origin by the request's own Host: a browser sets Host from the
        // URL it is fetching, so `Origin == Host` can only come from a page
        // this server served. "The request's own Host" is the front door's,
        // not the internal one a proxy hop rewrote it to.
        let front_door = effective_host(host_header, forwarded_host, peer, &self.trusted_proxies);
        match normalize_origin(front_door.as_deref()) {
            Some(own) => same_site(origin, &own),
            None => false,
        }
    }

    /// Allow or deny one request. Pure: no I/O, no app state.
    pub fn evaluate(&self, request: &CsrfRequest<'_>) -> CsrfDecision {
        let method = request.method.to_ascii_uppercase();
        if SAFE_METHODS.contains(&method.as_str()) {
            return allow("safe-method");
        }
        if request
            .authorization
            .unwrap_or("")
            .trim()
            .to_ascii_lowercase()
            .starts_with("bearer ")
        {
            // Not ambient: a cross-site page cannot attach this header.
            return allow("bearer-auth");
        }
        if !has_session_cookie(request.cookie_header) {
            return allow("no-session-cookie");
        }

        let mut stated = normalize_origin(request.origin);
        if stated.is_none() && request.origin.is_some_and(|value| !value.is_empty()) {
            // An opaque origin ("null") explicitly claims untrusted provenance.
            return deny("opaque-origin");
        }
        if stated.is_none() {
            stated = normalize_origin(request.referer);
            if stated.is_none() {
                return if self.bind_is_loopback {
                    // Non-browser client on the same machine (CLI, desktop
                    // shell, curl). Nothing on the network can reach this bind.
                    allow("no-origin-loopback-bind")
                } else {
                    deny("no-origin-reachable-bind")
                };
            }
        }

        let stated = stated.expect("checked above");
        if self.origin_is_trusted(&stated, request.host, request.forwarded_host, request.peer) {
            allow("same-site-or-trusted-origin")
        } else {
            deny("cross-site-origin")
        }
    }

    /// Evaluate straight off an axum request's parts.
    pub fn evaluate_headers(
        &self,
        method: &Method,
        headers: &HeaderMap,
        peer: Option<&str>,
    ) -> CsrfDecision {
        let read = |name: &str| headers.get(name).and_then(|value| value.to_str().ok());
        self.evaluate(&CsrfRequest {
            method: method.as_str(),
            origin: read("origin"),
            referer: read("referer"),
            host: read("host"),
            cookie_header: read("cookie"),
            authorization: read("authorization"),
            forwarded_host: read("x-forwarded-host"),
            peer,
        })
    }
}

/// The server's own origin plus loopback, in both schemes.
fn default_origins(server_host: &str, server_port: u16) -> Vec<String> {
    let hosts = [server_host, "localhost", "127.0.0.1", "[::1]"];
    let mut origins = Vec::new();
    for host in hosts {
        if host.is_empty() {
            continue;
        }
        for scheme in ["http", "https"] {
            origins.push(format!("{scheme}://{host}:{server_port}"));
        }
    }
    origins
}

fn allow(reason: &'static str) -> CsrfDecision {
    CsrfDecision {
        allowed: true,
        reason,
    }
}

fn deny(reason: &'static str) -> CsrfDecision {
    CsrfDecision {
        allowed: false,
        reason,
    }
}

/// The 403 body, written the way `core/csrf.py` writes it: `json.dumps` with
/// Python's default separators, so the spaces after `:` and `,` are real.
pub fn csrf_denied_response(reason: &str) -> Response {
    let body = dumps_spaced(&[
        ("detail", json!(DENIED_DETAIL)),
        ("error", json!("csrf_origin_rejected")),
        ("reason", json!(reason)),
    ]);
    json_response_utf8(StatusCode::FORBIDDEN, &body)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn policy() -> CsrfOriginPolicy {
        CsrfOriginPolicy::new(&[], "127.0.0.1", 4825, true, Vec::new())
    }

    fn request<'a>(method: &'a str) -> CsrfRequest<'a> {
        CsrfRequest {
            method,
            host: Some("127.0.0.1:4825"),
            ..CsrfRequest::default()
        }
    }

    #[test]
    fn safe_methods_are_never_guarded() {
        for method in ["GET", "head", "OPTIONS", "trace"] {
            let decision = policy().evaluate(&request(method));
            assert_eq!(decision, allow("safe-method"), "{method}");
        }
    }

    #[test]
    fn bearer_auth_is_exempt_before_the_cookie_check() {
        let mut probe = request("POST");
        probe.authorization = Some("  Bearer abc  ");
        probe.cookie_header = Some("session_token=x");
        probe.origin = Some("https://evil.example");
        assert_eq!(policy().evaluate(&probe), allow("bearer-auth"));
    }

    #[test]
    fn no_session_cookie_is_nothing_to_forge() {
        let mut probe = request("POST");
        probe.origin = Some("https://evil.example");
        assert_eq!(policy().evaluate(&probe), allow("no-session-cookie"));
        probe.cookie_header = Some("other=1");
        assert_eq!(policy().evaluate(&probe), allow("no-session-cookie"));
    }

    #[test]
    fn an_opaque_origin_is_refused() {
        let mut probe = request("POST");
        probe.cookie_header = Some("session_token=x");
        probe.origin = Some("null");
        assert_eq!(policy().evaluate(&probe), deny("opaque-origin"));
    }

    #[test]
    fn no_origin_depends_on_the_bind() {
        let mut probe = request("POST");
        probe.cookie_header = Some("session_token=x");
        assert_eq!(policy().evaluate(&probe), allow("no-origin-loopback-bind"));

        let reachable = CsrfOriginPolicy::new(&[], "0.0.0.0", 4825, false, Vec::new());
        assert_eq!(reachable.evaluate(&probe), deny("no-origin-reachable-bind"));
    }

    #[test]
    fn referer_stands_in_for_a_missing_origin() {
        let mut probe = request("POST");
        probe.cookie_header = Some("session_token=x");
        probe.referer = Some("http://127.0.0.1:4825/app");
        assert_eq!(
            policy().evaluate(&probe),
            allow("same-site-or-trusted-origin")
        );
        probe.referer = Some("https://evil.example/x");
        assert_eq!(policy().evaluate(&probe), deny("cross-site-origin"));
    }

    #[test]
    fn a_cross_site_origin_is_refused_and_a_trusted_one_is_not() {
        let mut probe = request("POST");
        probe.cookie_header = Some("session_token=x");
        probe.origin = Some("https://evil.example");
        assert_eq!(policy().evaluate(&probe), deny("cross-site-origin"));

        let trusted = CsrfOriginPolicy::new(
            &["https://evil.example".to_string()],
            "127.0.0.1",
            4825,
            true,
            Vec::new(),
        );
        assert_eq!(
            trusted.evaluate(&probe),
            allow("same-site-or-trusted-origin")
        );
    }

    #[test]
    fn origin_equal_to_the_host_is_same_site() {
        let mut probe = CsrfRequest {
            method: "POST",
            host: Some("brain.example.org"),
            cookie_header: Some("session_token=x"),
            origin: Some("https://brain.example.org"),
            ..CsrfRequest::default()
        };
        assert_eq!(
            policy().evaluate(&probe),
            allow("same-site-or-trusted-origin")
        );
        // A proxy that rewrote Host is why the forwarded authority is resolved
        // first — but only from a peer that may forward.
        probe.host = Some("127.0.0.1:4826");
        probe.origin = Some("https://front.door");
        assert_eq!(policy().evaluate(&probe), deny("cross-site-origin"));
        probe.forwarded_host = Some("front.door");
        probe.peer = Some("8.8.8.8");
        assert_eq!(policy().evaluate(&probe), deny("cross-site-origin"));
        probe.peer = Some("127.0.0.1");
        assert_eq!(
            policy().evaluate(&probe),
            allow("same-site-or-trusted-origin")
        );
    }

    #[test]
    fn a_hostless_request_cannot_be_same_site() {
        let probe = CsrfRequest {
            method: "POST",
            cookie_header: Some("session_token=x"),
            origin: Some("https://brain.example.org"),
            ..CsrfRequest::default()
        };
        assert_eq!(policy().evaluate(&probe), deny("cross-site-origin"));
    }

    #[test]
    fn the_allowlist_is_deduplicated_and_normalised() {
        let policy = CsrfOriginPolicy::new(
            &[
                "http://127.0.0.1:4825".to_string(),
                "  ".to_string(),
                "https://Brain.Example.ORG:443".to_string(),
            ],
            "127.0.0.1",
            4825,
            true,
            Vec::new(),
        );
        let trusted = policy.trusted_origins();
        assert!(trusted.contains(&("https".into(), "brain.example.org".into(), None)));
        let loopback_http = trusted
            .iter()
            .filter(|origin| origin.1 == "127.0.0.1" && origin.0 == "http")
            .count();
        assert_eq!(loopback_http, 1);
    }

    #[test]
    fn the_refusal_body_carries_python_spacing() {
        let response = csrf_denied_response("cross-site-origin");
        assert_eq!(response.status(), StatusCode::FORBIDDEN);
        assert_eq!(
            response.headers().get("content-type").unwrap(),
            "application/json; charset=utf-8"
        );
    }

    #[test]
    fn header_evaluation_matches_the_pure_policy() {
        let mut headers = HeaderMap::new();
        headers.insert("cookie", "session_token=x".parse().unwrap());
        headers.insert("origin", "https://evil.example".parse().unwrap());
        headers.insert("host", "127.0.0.1:4825".parse().unwrap());
        let decision = policy().evaluate_headers(&Method::POST, &headers, Some("127.0.0.1"));
        assert_eq!(decision, deny("cross-site-origin"));
    }
}
